### ISSUE TITLE
http-client: $select as params object

### ISSUE THREAD

**Issue by vobu** [SAP/cloud-sdk-js#2561]
Hi,

according [to the docs](https://sap.github.io/cloud-sdk/docs/js/features/connectivity/generic-http-client), utilizing `params` from https://axios-http.com/docs/req_config should be possible. 
But doing (code shortened for brevity) the below to forward an OData `$select` to axios unfortunately fails:
```js
const params = new URLSearchParams()
params.append("$select", "someProperty")
//...
requestConfig["params"] = params
const result = await executeHttpRequest(destinationOptions, requestConfig, httpRequestOptions)
```
"fails" means that the `params` object never reaches the axios instance (via `requestConfig`) intact.
Culprit seems to be `http-client.js`, specifically `buildRequestWithMergedHeadersAndQueryParameters`:
`const encodedParameters = getEncodedParameters(parameters, requestConfig);` wrangles the above `params` object into:
```js
{
  $select: {"s","o","m","e","P","r","o","p","e","r","t","y"}
}
```
Subsequently, the generic http client throws that `$select` key away in its utilization of axios.
Can you adivse/doublecheck how to correctly use the `params` option in the generic http client, please?
thx!

**Comment by vobu** [comment]
yeah, well:
```shell
@vobu ➜ /workspaces/cloud-sdk-js (9a4707e9) $ git diff v2.2.0 v2.5.0 -- packages/http-client/src/http-client.ts | grep buildRequestWithMergedHeadersAndQueryParameters
```
don't indicate any change btw `2.2.0` and `2.5.0`
any tests i can look at showing/testing axios-compatible `params`?

**Comment by jjtang1985** [comment]
Hi @vobu ,

I guess you should not use `URLSearchParams` for building the parameters, instead, you should use an object for the `params`
I'll give you one example in your case:
```
params =  {
    "$select": 'someProperty'
 }
requestConfig.params = params
```

Let me know, whether it works.

In general, in addition to the http client, SDK also provides convenient OData Clients. [doc](https://sap.github.io/cloud-sdk/docs/js/features/odata/overview)
You can find an example [here](https://sap.github.io/cloud-sdk/docs/js/features/odata/use-odata-v2-type-safe-client-for-javascript-typescript#select), which shows how you can build an OData request with `$select`.
The type safe client does the code automation for you and you don't even have to remember all the properties of an entity .

If you are using SAP S/4HANA Cloud or On-Prem system, you can fine our pre-generated clients. [doc](https://sap.github.io/cloud-sdk/docs/js/features/odata/overview#pregenerated-odata-clients)

If not, you can also generate your own OData client and here is the [doc](https://sap.github.io/cloud-sdk/docs/js/features/odata/generate-odata-client).

Let me you know, if you have questions regarding the clients.

Best regards,
Junjie

**Comment by jjtang1985** [comment]
@marikaner , I just realised that [axois](https://github.com/axios/axios) supports using both plain object and `URLSearchParams` (node standards) as `param` of the request config.
Perhaps we should also consider the alignment.

**Comment by vobu** [comment]
hi @jjtang1985 ,

thanks for the explanations.
this...

> ```
> params =  {
>     "$select": 'someProperty'
>  }
> requestConfig.params = params
> ```
> 
> Let me know, whether it works.

we tried, and it doesn't work - be it a plain JS or a `URLSearchParam` object.

**Comment by jjtang1985** [comment]
Hi @vobu,

>  be it a plain JS or a URLSearchParam object.

Currently, the SDK only supports using JS object, although `axois` supports both plain JS object and `URLSearchParam`.

If my previous does not work, would you like to share the reason?
Is it the same reason as before, i.e., the `$select` key was thrown?
Would you like to test the example below, so we understand whether the `$` is a problem?
```
params =  {
    'testKey': 'testValue',
    "$select": 'someProperty'
 }
requestConfig.params = params
```

**Comment by jjtang1985** [comment]
@vobu ,

Also, I assume you updated to `2.5.0`, as we fixed some issues for `executeHttpReqeust`.

**Comment by vobu** [comment]
> Hi @vobu,
> 
> > be it a plain JS or a URLSearchParam object.
> 
> Currently, the SDK only supports using JS object, although `axois` supports both plain JS object and `URLSearchParam`.
> 
> If my previous does not work, would you like to share the reason? Is it the same reason as before, i.e., the `$select` key was thrown? Would you like to test the example below, so we understand whether the `$` is a problem?
> 
> ```
> params =  {
>     'testKey': 'testValue',
>     "$select": 'someProperty'
>  }
> requestConfig.params = params
> ```

i did exactly that during my tries, and with `2.2.0` this also resulted in an empty axios params object.

> Could you please share, whether you tried 2.5.0 and the result?

as the part of the app utilizing the cloud-sdk is not covered sufficiently by unit tests yet, and b/c of more pressing project work, I'm hesitant of updating the dependency and trying stuff manually.
this will happen eventually though (both the increase in test coverage and the upgrade), then i can get back to you.
still, i would hope that i've provided enough information that you could add a test case in the cloud-sdk itself to cover for the use-case of providing `$select` as a property to the `executeHttpRequest`'s `httpRequestOptions` parameter.

**Comment by jjtang1985** [comment]
Hi @vobu ,

> i did exactly that during my tries, and with 2.2.0 this also resulted in an empty axios params object.

I tested the coded example (that I shared [before](https://github.com/SAP/cloud-sdk-js/issues/2561#issuecomment-1154379997)) with both version `2.2.0` and `2.5.0` and here is the result:
- `2.5.0` worked fine
- `2.2.0` did not work as we introduced some breaking changes, which was fixed in `2.5.0`.

>  I'm hesitant of updating the dependency and trying stuff manually.

I understand your situation and here are some suggestions:
1. upgrade to `2.5.0`, where the bug was fixed
2. still use `2.2.0`, but you have to adjust the code (see example below) when calling `executeHttpRequest`, this is not recommended, because you might want to revert your code adjustment again when upgrading to `2.5.0`
3. use our OData client instead of http client as mentioned [above](https://github.com/SAP/cloud-sdk-js/issues/2561#issuecomment-1154379997)

Here is the example that works with `2.2.0`, where you have to use `requestConfig` as an additional key.
```ts
params =  {
    // additional key you need
    requestConfig:{
        "$select": 'someProperty'
    }
 }
requestConfig.params = params
```

I hope this clarifies a bit more.
Let me know, if you have questions.

**Comment by marikaner** [comment]
@vobu As this bug is fixed and tested, this issue is resolved. In case you still encounter problems with the latest version, please reach out again.
